packages/core/src/utils/objects/deepClone.ts

```diff
@@ -35,37 +41,45 @@ export function deepClone(source: any, stack = new WeakMap()): any {
     return new RegExp(source);
   }
 
-  const stacked = stack.get(source);
+  if (Array.isArray(source)) {
+    const clone: unknown[] = [];
+    stack.set(source, clone);
+    source.forEach((item, idx) => (clone[idx] = deepClone(item, stack)));
+    return clone;
+  }
 
-  if (stacked) {
-    // See issue #1619
-    return stacked;
+  if (source instanceof Map) {
+    const clone = new Map();
+    stack.set(source, clone);
+    source.forEach((value, key) => clone.set(deepClone(key, stack), deepClone(value, stack)));
+    return clone;
   }
 
-  if (isArray(source)) {
-    dest = [];
-  } else {
-    dest = {};
-    stack.set(source, dest);
+  if (source instanceof Set) {
+    const clone = new Set();
+    stack.set(source, clone);
+    source.forEach((value) => clone.add(deepClone(value, stack)));
+    return clone;
   }
 
-  for (const key in source) {
-    // Use getOwnPropertyDescriptor instead of source[key] to prevent from triggering setter/getter.
-    const descriptor = Object.getOwnPropertyDescriptor(source, key)!;
+  const clone = Object.create(Reflect.getPrototypeOf(source));
+  stack.set(source, clone);
+
+  Reflect.ownKeys(source).forEach((key) => {
+    // respects property descriptors and the prototype chain more explicitly, which is important for objects with getter/setter.
```

Comment:
// use property descriptors to prevent from triggering setter/getter.
```suggestion
    // use property descriptors to prevent from triggering setter/getter.
```